### ISSUE TITLE
Add explicit export lists to modules

### ISSUE THREAD

**Issue by rwe** [agda/agda#4824]
Derived from #3512, and specifically [this note](https://github.com/agda/agda/issues/3512#issuecomment-667247235), replicated here:

Beyond build + rebuild times, and regardless of improvements to the actual inter-module coupling, the "everything open-exports and open-imports" style is likely introducing unnecessary *runtime* costs.

[From the GHC wiki](https://wiki.haskell.org/Performance/GHC):
> For best results, use an _explicit export list_. If you do, GHC can inline any non-exported functions that are only called once, even if they are very big. Without an explicit export list, GHC must assume that every function is exported, and hence (to avoid code bloat) is more conservative about inlining.

This is potentially exacerbated by the fact that `Agda.cabal` defines nearly all modules as public, rather than under `other-modules`, meaning that even if GHC _did_ do whole-component analysis, it would be prevented from doing so.

Note that **explicit _import_ lists are not proposed by this issue.** There's concern that requiring explicit imports will interfere with local development; it's a much harder pill to swallow. However explicit export lists are easier to deal with.

**Comment by jespercockx** [comment]
Note that we are currently using two non-standard flags that might mean the general Haskell wisdom does not apply here:

https://github.com/agda/agda/blob/df5b9410d6b7c4fd70db9d3ebcad6e93fa847681/Agda.cabal#L612-L613

**Comment by rwe** [comment]
That's good to know, I didn't realize the connection with those flags. So perhaps no runtime performance to be gained.

Nonetheless I think this might be a good idea anyway (even approached incrementally) both for refactoring work and also as part making "public interface" a slightly more visible consideration at the module level (#4826). Anyone object to the issue floating for those reasons?

As a total aside, I do wish Haskell supported something like a `export foo :: Int` syntax à la modern JS and some other languages, versus bundling identifiers at the top. On the one hand it's nice to get a global view of the public interface; on the other it can be a bit tedious jumping between module headers and bodies.

**Comment by andreasabel** [comment]
My main problem with export lists was compilation time due to a not very smart recompilation machinery in GHC: After adding a function f in module A and using it in module B, it used to take up to 15 min compilation time to the compilation error where I realized that I forgot to add f to the export list in A.  Going back and changing A by adding f to the export list would then trigger recompilation of everything again.  This problem is now _mostly_ solved by the `make type-check` quick compilation.  However, a slight annoyance-factor and productivity cost will remain by having to maintain explicit import lists.
I does make a lot of sense for well-designed APIs, but I am not sure we have those. 
Export lists seem to me a bureaucracy that suitable tooling could take off my shoulders.  But I am not aware there is such tooling for Haskell.

**Comment by rwe** [comment]
@andreasabel I agree that the export-list bureaucracy is an annoyance with Haskell. The requirement of one identifier per export versus a keyword per export like many other languages (`export …`, `public …`, etc) really encourages some counterproductive "all hangin' out" defaults. And unfortunately it's not an easily toolable problem because, unlike the minimal-import lists you can get from GHC, exports are a cross-module global-view issue. ("Is this symbol used anywhere by anything in the world?").

Regarding:
> makes sense for well-designed APIs, but I am not sure we have those

I agree-ish, though they're not so far off, and I also think it's hard to get there without taking at least some small steps.

Re the turnaround time, which used to be a bigger pain for me than it is now: if you haven't already, you might want to try the [haskell language server](https://github.com/haskell/haskell-language-server). I've been using it recently with GHC 8.8.3 and it's been working very well and doesn't tend to require frequent rebuilds. So in the "I forgot to add…" scenario, it'll notice that immediately as you're typing. (I use VSCode though [it supports Emacs](https://github.com/haskell/haskell-language-server#using-haskell-language-server-with-emacs) and others as well).

**Comment by andreasabel** [comment]
@rwe : Let me emphasize/clarify that I am ok with explicit export lists, but I will probably not invest my time to push this.
